_posts/2018-08-26-eager-dcgan/eager-dcgan.Rmd

```diff
@@ -0,0 +1,460 @@
+          self$batchnorm1(training = training) %>%
+          self$leaky_relu1() %>%
+          k_reshape(shape = c(-1, 7, 7, 64)) %>%
+          self$conv1() %>%
+          self$batchnorm2(training = training) %>%
+          self$leaky_relu2() %>%
+          self$conv2() %>%
+          self$batchnorm3(training = training) %>%
+          self$leaky_relu3() %>%
+          self$conv3()
+      }
+    })
+  }
+
+```
+
+
+## Discriminator
+
+The discriminator is just a pretty normal convolutional network outputting a score. Here, usage of "score" instead of "probability" is on purpose: If you look at the last layer, it is fully connected, of size 1 but lacking the usual sigmoid activation. This is because unlike Keras' `loss_binary_crossentropy`, the loss function we'll be using here - `tf$losses$sigmoid_cross_entropy` - works with the raw logits, not the outputs of the sigmoid.
+
+```{r}
+discriminator <-
+  function(name = NULL) {
+    keras_model_custom(name = name, function(self) {
+      
+      self$conv1 <- layer_conv_2d(
+        filters = 64,
+        kernel_size = c(5, 5),
+        strides = c(2, 2),
+        padding = "same"
+      )
+      self$leaky_relu1 <- layer_activation_leaky_relu()
+      self$dropout <- layer_dropout(rate = 0.3)
+      self$conv2 <-
+        layer_conv_2d(
+          filters = 128,
+          kernel_size = c(5, 5),
+          strides = c(2, 2),
+          padding = "same"
+        )
+      self$leaky_relu2 <- layer_activation_leaky_relu()
+      self$flatten <- layer_flatten()
+      self$fc1 <- layer_dense(units = 1)
+      
+      function(inputs,
+               mask = NULL,
+               training = TRUE) {
+        inputs %>% self$conv1() %>%
+          self$leaky_relu1() %>%
+          self$dropout(training = training) %>%
+          self$conv2() %>%
+          self$leaky_relu2() %>%
+          self$flatten() %>%
+          self$fc1()
+      }
+    })
+  }
+```
+
+
+## Setting the scene
+
+Before we can start training, we need to create the usual components of a deep learning setup: the model (or models, in this case), the loss function(s), and the optimizer(s).
+
+Model creation is just a function call, with a little extra on top:
+
+```{r}
+generator <- generator()
+discriminator <- discriminator()
+
+# https://www.tensorflow.org/api_docs/python/tf/contrib/eager/defun
+generator$call = tf$contrib$eager$defun(generator$call)
+discriminator$call = tf$contrib$eager$defun(discriminator$call)
+```
+
+[_defun_](https://www.tensorflow.org/api_docs/python/tf/contrib/eager/defun) compiles a Python function (once per different combination of argument shapes and non-tensor objects values)) into a TensorFlow graph, and is used to speed up computations. This comes with side effects and possibly unexpected behavior - please consult the documentation for the details. Here, we were mainly curious in how much of a speedup we might notice when using this from R - in our example, it resulted in a speedup of 130%.
```

Comment:
compiles a Python or R function ??? (users might not appreciate that R functions are in fact manifest to TF as Python functions)

Reply:
changing to R